### ISSUE TITLE
[New Component] WooCommerce - Initial Research

### ISSUE THREAD

**Issue by vladolaru** [pixelgrade/components#20]
In order to keep the components ahead of the curve, I propose we start investigating what would a WooCommerce component look like.

_Notice: Due to the nature of the problem at hand, it could be a quite technical discussion. I will try and keep it as inclusive as possible. Just ask for clarifications and I will do my best to help shed some light._

## Driving Principles
 
1. Using **only WooCommerce hooks** at the component level (if one wants to overwrite WooCommerce templates it can do it at the theme level)
2. **Using an approach as light as possible** vis-a-vis WooCommerce's way of doing things, thus avoiding as much as possible future issues with new WooCommerce versions
3. **Integrating existing components** with the WooCommerce templates (think header, footer, featured image)
4. What would be the **scope of the component** in terms behavior and templating
5. Establishing a **configuration starting point** flexible enough to avoid templates overwrites
6. To be determined after receiving feedback...

## Possible Difficulties

- **WooCommerce is standardizing and changing** the way it handles things migrating towards a more theme-friendly default behavior; this might help us or it might restrict us too much (being too opinionated) - not sure if this is an issue
- **The sheer complexity of the WooCommerce hook system** and the complex ways things interact
- Drawing a line between **keeping things manageable for us versus allowing users** (and designers) to access the full potential of WooCommerce
- The WooCommerce **hooks system might be too restrictive** for our needs and we might have troubles playing nice with it
- The WooCommerce **templating system might be difficult to interact with at a lower level** (needed for the reusable, predictable manner in which we wish the components to behave), making us prone to update difficulties

## Next Steps

- [ ] Properly understand _the philosophy_ behind WooCommerce's frontend and the way it is influenced by the backend
- [ ] Analyze _the Storefront theme_ because that is the "official" way of doing things
- [ ] Analyze and classify the WooCommerce _frontend hooks_ and group them in a way convenient for the components system (make complexity manageable)
- [ ] _Prioritize_ the WooCommerce frontend features by importance and likelihood of use by users
- [ ] Define the component's scope according to all of the above (maybe we settle on multiple components, not a single, massive one)
- [ ] Devise a game plan according to the scope and prioritization above
- [ ] To be determined after receiving feedback...

---

These are only my initial thoughts and they are very likely to change after progressing, researching, looking at different approaches, and hearing from others. But we need to start somewhere, so let's get the ball rolling and think about it.

_I am giving a general ping @pixelgrade/everybody, because there is no need to narrow the field right from the getgo._

**Comment by georgeolaru** [comment]
A short reinforcement on using hooks instead of editing templates: [Template File Guidelines for Devs and Theme Authors](https://github.com/woocommerce/woocommerce/wiki/Template-File-Guidelines-for-Devs-and-Theme-Authors)

**Comment by raduconst** [comment]
@vladolaru I agree with creating a WooCommerce component and the use of hooks. We know that we have often confronted with the problem of "outdated templates" and I think that a more simplistic approach (without files changes) is easier to maintain in the long run.

If I can give a helping hand for implementation or testing, I'm right here.
